File: src/main/java/de/kxmpetentes/engine/model/EmbedModel.java
package de.kxmpetentes.engine.model;

import de.kxmpetentes.engine.DiscordCore;
import net.dv8tion.jda.api.EmbedBuilder;
import net.dv8tion.jda.api.entities.Guild;
import net.dv8tion.jda.api.entities.Message;
import net.dv8tion.jda.api.entities.PrivateChannel;
import net.dv8tion.jda.api.entities.TextChannel;

import java.awt.*;
import java.time.OffsetDateTime;
import java.time.temporal.TemporalAccessor;

/**
 * @author kxmpetentes
 * Website: kxmpetentes.de
 * GitLab: gitlab.com/kxmpetentes
 * GitHub: git.kxmpetentes.de
 * Erstellt am: 04.01.2021 um 22:00
 */

public class EmbedModel {

    private String author;
    private String title;
    private String image;
    private String thumbnail;
    private String footer;
    private String footerIcon;
    private Color color;
    private String text;
    private EmbedBuilder embedBuilder = new EmbedBuilder();
    private TemporalAccessor timestep;
    private String authorLink;

    public EmbedModel(String text) {
        this.text = text;
    }

    public EmbedModel(String title, String author, String image, Color color, String text, String thumbnail, String footer) {
        this.title = title;
        this.author = author;
        this.image = image;
        this.color = color;
        this.text = text;
        this.thumbnail = thumbnail;
        this.timestep = OffsetDateTime.now();
        this.footer = footer;
    }

    public EmbedModel createMessage() {
        embedBuilder.setTitle(title);
        embedBuilder.setAuthor(author, authorLink);
        embedBuilder.setImage(image);
        embedBuilder.setColor(color);
        embedBuilder.setDescription(text);
        embedBuilder.setFooter(footer, footerIcon);
        embedBuilder.setThumbnail(thumbnail);
        embedBuilder.setTimestamp(timestep);

        return this;
    }

    public EmbedModel addField(String name, String value, boolean inline) {
        embedBuilder.addField(name, value, inline);

        return this;
    }

    public void sendPrivate(PrivateChannel privateChannel) {
        privateChannel.sendMessage(this.embedBuilder.build()).queue();
    }

    public void sendTextChannel(TextChannel channel) {
        channel.sendMessage(this.embedBuilder.build()).queue();
    }

    public void reply(Message message) {
        message.reply(embedBuilder.build()).queue();
    }

    public void reply(Message message, EmbedBuilder embedBuilder) {
        message.reply(embedBuilder.build()).queue();
    }

    public void setAuthor(String author) {
        this.author = author;
    }

    public void setAuthorLink(String authorLink) {
        this.authorLink = authorLink;
    }

    public void setColor(Color color) {
        this.color = color;
    }

    public EmbedBuilder getEmbedBuilder() {
        return embedBuilder;
    }

    public void setEmbedBuilder(EmbedBuilder embedBuilder) {
        this.embedBuilder = embedBuilder;
    }

    public void setFooter(String footer) {
        this.footer = footer;
    }

    public void setImage(String image) {
        this.image = image;
    }

    public void setText(String text) {
        this.text = text;
    }

    public void setThumbnail(String thumbnail) {
        this.thumbnail = thumbnail;
    }

    public void setTitle(String title) {
        this.title = title;
    }

    public void setTimestep(TemporalAccessor timestep) {
        this.timestep = timestep;
    }

    public void setFooterIcon(String footerIcon) {
        this.footerIcon = footerIcon;
    }

    public void setFooterIcon() {
        this.footerIcon = DiscordCore.getInstance().getBotIconURL();
    }

}

File: src/main/java/de/kxmpetentes/engine/model/EmbedModel.kt
package de.kxmpetentes.engine.model

import net.dv8tion.jda.api.EmbedBuilder
import net.dv8tion.jda.api.entities.Message
import net.dv8tion.jda.api.entities.PrivateChannel
import net.dv8tion.jda.api.entities.TextChannel
import org.jetbrains.annotations.Nullable
import java.awt.Color
import java.time.Duration
import java.time.OffsetDateTime

/**
 * @param title         the title of the Embed
 * @param author        the author of the Embed
 * @param image         the image of the Embed
 * @param color         the color of the Embed
 * @param text          the text to be in the Embed
 * @param thumbnail     the thumbnail right in the top right corner
 * @param footer        the footer of the Embed
 */
data class EmbedModel(
    @Nullable var title: String,
    @Nullable var author: String,
    @Nullable var image: String,
    @Nullable var color: Color,
    @Nullable var text: String,
    @Nullable var thumbnail: String,
    @Nullable var footer: String,
) {

    private var embedBuilder: EmbedBuilder = EmbedBuilder()
    private lateinit var authorLink: String
    private lateinit var description: String
    private lateinit var footerIcon: String
    private var timestamp: OffsetDateTime = OffsetDateTime.now()

    /**
     * creates the Message (must be run, if the embed should look correctly)
     */
    fun createMessage(): EmbedModel {

        embedBuilder.setTitle(title)
            .setAuthor(author, authorLink)
            .setColor(color)
            .setImage(image)
            .setDescription(description)
            .setFooter(footer, footerIcon)
            .setThumbnail(thumbnail)
            .setTimestamp(timestamp)
        return this
    }

    /**
     * @param name          the name of the field
     * @param value         the value of the field
     * @param inline        should the field be inline with other fields
     */
    fun addField(name: String, value: String, inline: Boolean): EmbedModel {
        embedBuilder.addField(name, value, inline)

        return this
    }

    /**
     * sends the embed to a private channel
     */
    fun sendPrivate(privateChannel: PrivateChannel) {
        privateChannel.sendMessage(embedBuilder.build()).queue()
    }

    /**
     * sends the embed to a text channel
     */
    fun sendToTextChannel(textChannel: TextChannel) {
        textChannel.sendMessage(embedBuilder.build()).queue()
    }

    /**
     * sends the embed to a text channel and deletes it after a given delay
     */
    fun sendToTextChannelAndDeleteLater(textChannel: TextChannel, delay: Long) {
        textChannel.sendMessage(embedBuilder.build()).delay(Duration.ofSeconds(delay)).flatMap(Message::delete).queue()
    }

    /**
     * replies to a message
     */
    fun reply(message: Message, mention: Boolean) {
        message.reply(embedBuilder.build()).mentionRepliedUser(mention).queue()
    }

    /**
     * replies to a message and deletes it after a given delay
     */
    fun replyAndDeleteLater(message: Message, delay: Long) {
        message.reply(embedBuilder.build()).delay(Duration.ofSeconds(delay)).flatMap(Message::delete).queue()
    }

}
